File: projects/Python-Playwright-Automation/framework/pages/automation_practice_page.py
"""Page Object for the public AutomationPractice page.

Locator policy for this file
----------------------------
The page is a third-party practice site, so its markup can change without notice.
Where a stable, semantic hook exists we use a user-facing locator (`get_by_role`,
`get_by_label`, header-text filtering). Where the site only offers an id, we use the
id -- but we keep every one of them in THIS file, so a markup change is a one-file fix
and never touches a test.

Run `python scripts/verify_selectors.py` to check every locator below in one pass.
"""
from playwright.sync_api import FrameLocator, Locator, Page

from framework.pages.base_page import BasePage


class AutomationPracticePage(BasePage):
    def __init__(self, page: Page):
        super().__init__(page)

        # --- Radio buttons ---
        self.radio_buttons = page.locator("input[name='radioButton']")

        # --- Suggestion / autocomplete ---
        self.autocomplete_input = page.locator("#autocomplete")

        # --- Static dropdown ---
        self.dropdown = page.locator("#dropdown-class-example")

        # --- Checkboxes ---
        self.checkbox_1 = page.locator("#checkBoxOption1")
        self.checkbox_2 = page.locator("#checkBoxOption2")
        self.checkbox_3 = page.locator("#checkBoxOption3")

        # --- Alerts ---
        self.alert_name_input = page.locator("#name")
        self.alert_button = page.locator("#alertbtn")
        self.confirm_button = page.locator("#confirmbtn")

        # --- New tab / new window ---
        self.open_tab_link = page.locator("#opentab")
        self.open_window_button = page.locator("#openwindow")

        # --- Show / hide ---
        self.displayed_text = page.locator("#displayed-text")
        self.hide_button = page.locator("#hide-textbox")
        self.show_button = page.locator("#show-textbox")

        # --- Mouse hover ---
        self.mouse_hover_button = page.locator("#mousehover")

        # --- iFrame ---
        self.iframe_selector = "#courses-iframe"

    # ------------------------------------------------------------------ #
    # Resilient table locators: find the table by its HEADER TEXT rather  #
    # than by a brittle class/id, so a restyle doesn't break them.        #
    # ------------------------------------------------------------------ #
    @property
    def courses_table(self) -> Locator:
        """The Instructor / Course / Price table."""
        return self.page.locator("table").filter(has_text="Instructor").first

    @property
    def amount_table(self) -> Locator:
        """The fixed-header table containing an Amount column."""
        return self.page.locator("table").filter(has_text="Amount").first

    # ---------------------------- actions ----------------------------- #
    def select_radio(self, index: int) -> None:
        """Select the nth radio button (0-based)."""
        self.radio_buttons.nth(index).check()

    def select_dropdown_by_value(self, value: str) -> None:
        self.dropdown.select_option(value)

    def check_all_checkboxes(self) -> None:
        for box in (self.checkbox_1, self.checkbox_2, self.checkbox_3):
            box.check()

    def trigger_alert(self, name: str) -> None:
        self.alert_name_input.fill(name)
        self.alert_button.click()

    def trigger_confirm(self, name: str) -> None:
        self.alert_name_input.fill(name)
        self.confirm_button.click()

    def hide_textbox(self) -> None:
        self.hide_button.click()

    def show_textbox(self) -> None:
        self.show_button.click()

    def hover_menu(self) -> None:
        self.mouse_hover_button.hover()

    def courses_frame(self) -> FrameLocator:
        return self.page.frame_locator(self.iframe_selector)

    # ------------------------------ state ----------------------------- #
    def course_prices(self) -> list[int]:
        """Every value in the Price column, as ints."""
        rows = self.courses_table.locator("tbody tr")
        prices: list[int] = []
        for i in range(rows.count()):
            cells = rows.nth(i).locator("td")
            if cells.count() >= 3:
                raw = (cells.nth(2).inner_text() or "").strip()
                if raw.isdigit():
                    prices.append(int(raw))
        return prices

    def course_names(self) -> list[str]:
        rows = self.courses_table.locator("tbody tr")
        names: list[str] = []
        for i in range(rows.count()):
            cells = rows.nth(i).locator("td")
            if cells.count() >= 2:
                names.append((cells.nth(1).inner_text() or "").strip())
        return names

File: projects/TypeScript-Playwright-Automation/framework/pages/AutomationPracticePage.ts
import type { FrameLocator, Locator, Page } from '@playwright/test';

import { BasePage } from './BasePage';

/**
 * Page Object for the public AutomationPractice page.
 *
 * Locator policy
 * --------------
 * Every selector this framework depends on lives in THIS file, so a markup change
 * is a one-file fix that never touches a test. Tables are located by header text
 * rather than by class/id, which survives a restyle.
 *
 * These selectors are the same set verified passing in the Python project.
 * Re-check any time with: npm run verify:selectors
 */
export class AutomationPracticePage extends BasePage {
  // --- Radio buttons ---
  readonly radioButtons: Locator;

  // --- Suggestion / autocomplete ---
  readonly autocompleteInput: Locator;

  // --- Static dropdown ---
  readonly dropdown: Locator;

  // --- Checkboxes ---
  readonly checkbox1: Locator;
  readonly checkbox2: Locator;
  readonly checkbox3: Locator;

  // --- Alerts ---
  readonly alertNameInput: Locator;
  readonly alertButton: Locator;
  readonly confirmButton: Locator;

  // --- New tab / new window ---
  readonly openTabLink: Locator;
  readonly openWindowButton: Locator;

  // --- Show / hide ---
  readonly displayedText: Locator;
  readonly hideButton: Locator;
  readonly showButton: Locator;

  // --- Mouse hover ---
  readonly mouseHoverButton: Locator;

  // --- iFrame ---
  readonly iframeSelector = '#courses-iframe';

  constructor(page: Page) {
    super(page);

    this.radioButtons = page.locator("input[name='radioButton']");
    this.autocompleteInput = page.locator('#autocomplete');
    this.dropdown = page.locator('#dropdown-class-example');

    this.checkbox1 = page.locator('#checkBoxOption1');
    this.checkbox2 = page.locator('#checkBoxOption2');
    this.checkbox3 = page.locator('#checkBoxOption3');

    this.alertNameInput = page.locator('#name');
    this.alertButton = page.locator('#alertbtn');
    this.confirmButton = page.locator('#confirmbtn');

    this.openTabLink = page.locator('#opentab');
    this.openWindowButton = page.locator('#openwindow');

    this.displayedText = page.locator('#displayed-text');
    this.hideButton = page.locator('#hide-textbox');
    this.showButton = page.locator('#show-textbox');

    this.mouseHoverButton = page.locator('#mousehover');
  }

  // ------------------------------------------------------------------ //
  // Resilient table locators: find the table by its HEADER TEXT rather  //
  // than a brittle class/id, so a restyle doesn't break them.           //
  // ------------------------------------------------------------------ //
  get coursesTable(): Locator {
    return this.page.locator('table').filter({ hasText: 'Instructor' }).first();
  }

  get amountTable(): Locator {
    return this.page.locator('table').filter({ hasText: 'Amount' }).first();
  }

  // ---------------------------- actions ----------------------------- //
  async selectRadio(index: number): Promise<void> {
    await this.radioButtons.nth(index).check();
  }

  async selectDropdownByValue(value: string): Promise<void> {
    await this.dropdown.selectOption(value);
  }

  async checkAllCheckboxes(): Promise<void> {
    for (const box of [this.checkbox1, this.checkbox2, this.checkbox3]) {
      await box.check();
    }
  }

  async triggerAlert(name: string): Promise<void> {
    await this.alertNameInput.fill(name);
    await this.alertButton.click();
  }

  async triggerConfirm(name: string): Promise<void> {
    await this.alertNameInput.fill(name);
    await this.confirmButton.click();
  }

  async hideTextbox(): Promise<void> {
    await this.hideButton.click();
  }

  async showTextbox(): Promise<void> {
    await this.showButton.click();
  }

  async hoverMenu(): Promise<void> {
    await this.mouseHoverButton.hover();
  }

  coursesFrame(): FrameLocator {
    return this.page.frameLocator(this.iframeSelector);
  }

  // ------------------------------ state ----------------------------- //
  /** Every value in the Price column, as numbers. */
  async coursePrices(): Promise<number[]> {
    const rows = this.coursesTable.locator('tbody tr');
    const prices: number[] = [];

    for (let i = 0; i < (await rows.count()); i++) {
      const cells = rows.nth(i).locator('td');
      if ((await cells.count()) >= 3) {
        const raw = (await cells.nth(2).innerText()).trim();
        if (/^\d+$/.test(raw)) prices.push(Number(raw));
      }
    }
    return prices;
  }

  async courseNames(): Promise<string[]> {
    const rows = this.coursesTable.locator('tbody tr');
    const names: string[] = [];

    for (let i = 0; i < (await rows.count()); i++) {
      const cells = rows.nth(i).locator('td');
      if ((await cells.count()) >= 2) {
        names.push((await cells.nth(1).innerText()).trim());
      }
    }
    return names;
  }
}
